### ISSUE TITLE
Tracks pauses constantly on first play

### ISSUE THREAD

**Issue by jessicah** [librespot-org/librespot#864]
Once cached, it's fine, but the first time playing a track, it always has pauses.

**Comment by roderickvd** [comment]
Well that's something that myself and most others can't reproduce, or we would be swamped with such issues.
Are you sure that your network connection is in order (so that it can buffer fast enough?)
Does it also happen with cache disabled?

**Comment by jessicah** [comment]
Actually, I noticed one error message on screen before disappearing of a buffer underrun. Am using the ALSA back-end. I don't have this with other apps, such as Kodi.

**Comment by roderickvd** [comment]
> And now it's gone from no buffer underruns to an underrun storm :-/ Also, both `ncspot` and `librespot` consistently use around 50% of a CPU core on my system, which seems excessive, is that normal?

No. Find the cause of this issue and you will likely find your solution. Your system is bogged and this is causing I/O underruns.

First to be sure, was this compiled with the `--release` flag?

Second the Alsa period size of 4096 seems very small, doesn't it @JasonLG1979?
What is your Alsa output device?

Please post:
1. the full command line you are launching `librespot` with
2. your Alsa configuration file
3. the full verbose log from the point you launched `librespot`

**Comment by jessicah** [comment]
Even with @JasonLG1979's config, I'm still getting playback issues (with release build), seems buffers are still small?

2021-10-15T00:12:00Z TRACE librespot_playback::player] == Starting sink ==
[2021-10-15T00:12:00Z TRACE librespot_playback::audio_backend::alsa] Frames per Buffer: 8192
[2021-10-15T00:12:00Z TRACE librespot_playback::audio_backend::alsa] Frames per Period: 4096
[2021-10-15T00:12:00Z TRACE librespot_playback::audio_backend::alsa] Period Buffer size in bytes: 32768
[2021-10-15T00:12:00Z TRACE librespot_playback::audio_backend::alsa] Period Buffer capacity: 32768

**Comment by kingosticks** [comment]
To be honest I am still not sure, the online alsa docs and the source code have conflicting comments! Probably safest to avoid using it if possible and go with buffer_time instead, if you have to specify it for some reason.

**Comment by roderickvd** [comment]
Any progress on this @jessicah?

**Comment by jessicah** [comment]
Sorry, I'll give the suggestions a go now, been moving stuff, and haven't had a chance to retest. It's an Intel NUC i3-4010U, audio via HDMI.

**Comment by jessicah** [comment]
Well, it seems it's a weird ALSA bug. Running `aplay -D hw:0,3 --verbose ~/test.wav` gives the same issues, so I don't understand how Kodi can play audio without any hiccups. I tried passing several parameters for buffer/period size/time, but they all seem to get ignored by `aplay`.

For reference, here is output from aply with the last suggested configuration:
```
jessica@homecloud:~/source/librespot$ aplay -D hw:0,3 --verbose ~/test.wav
Playing WAVE '/home/jessica/test.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Hardware PCM card 0 'HDA Intel HDMI' device 3 subdevice 0
Its setup is:
  stream       : PLAYBACK
  access       : RW_INTERLEAVED
  format       : S16_LE
  subformat    : STD
  channels     : 2
  rate         : 48000
  exact rate   : 48000 (48000/1)
  msbits       : 16
  buffer_size  : 16384
  period_size  : 4096
  period_time  : 85333
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 4096
  period_event : 0
  start_threshold  : 16384
  stop_threshold   : 16384
  silence_threshold: 0
  silence_size : 0
  boundary     : 4611686018427387904
  appl_ptr     : 0
  hw_ptr       : 0
```

And without specifying a device explicitly:
```
jessica@homecloud:~/source/librespot$ aplay --verbose ~/test.wav
Playing WAVE '/home/jessica/test.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Plug PCM: Linear conversion PCM (S32_LE)
Its setup is:
  stream       : PLAYBACK
  access       : RW_INTERLEAVED
  format       : S16_LE
  subformat    : STD
  channels     : 2
  rate         : 48000
  exact rate   : 48000 (48000/1)
  msbits       : 16
  buffer_size  : 8192
  period_size  : 4096
  period_time  : 85333
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 4096
  period_event : 0
  start_threshold  : 8192
  stop_threshold   : 8192
  silence_threshold: 0
  silence_size : 0
  boundary     : 4611686018427387904
Slave: Direct Stream Mixing PCM
Its setup is:
  stream       : PLAYBACK
  access       : MMAP_INTERLEAVED
  format       : S32_LE
  subformat    : STD
  channels     : 2
  rate         : 48000
  exact rate   : 48000 (48000/1)
  msbits       : 32
  buffer_size  : 8192
  period_size  : 4096
  period_time  : 85333
  tstamp_mode  : NONE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 4096
  period_event : 0
  start_threshold  : 8192
  stop_threshold   : 8192
  silence_threshold: 0
  silence_size : 0
  boundary     : 4611686018427387904
Hardware PCM card 0 'HDA Intel HDMI' device 3 subdevice 0
Its setup is:
  stream       : PLAYBACK
  access       : MMAP_INTERLEAVED
  format       : S32_LE
  subformat    : STD
  channels     : 2
  rate         : 48000
  exact rate   : 48000 (48000/1)
  msbits       : 32
  buffer_size  : 8192
  period_size  : 4096
  period_time  : 85333
  tstamp_mode  : ENABLE
  tstamp_type  : MONOTONIC
  period_step  : 1
  avail_min    : 4096
  period_event : 0
  start_threshold  : 1
  stop_threshold   : 4611686018427387904
  silence_threshold: 0
  silence_size : 4611686018427387904
  boundary     : 4611686018427387904
  appl_ptr     : 0
  hw_ptr       : 0
```

So I guess may as well close this? I'm not sure what else I can do, especially when Kodi doesn't have any issues :-/

**Comment by roderickvd** [comment]
One last thing you could check is version `0.1.6` and `0.2.0`. Those versions approached the Alsa buffers slightly different. If either one of those work, then it's worth a discussion whether we should go back to that -- although I'd prefer not to. Anyway before we get there, it'd be nice if you could report back on that.

**Comment by JasonLG1979** [comment]
> One last thing you could check is version `0.1.6` and `0.2.0`. Those versions approached the Alsa buffers slightly different. If either one of those work, then it's worth a discussion whether we should go back to that -- although I'd prefer not to. Anyway before we get there, it'd be nice if you could report back on that.

I think it may be a case of the device only accepting a very small set of a combination of buffer size, period size, number of periods parameters as even `aplay` is having trouble with it. As I've said it may just be a matter of experimenting with different values.

I'll see if I can't have a look at Kodi's source and see what they're doing.

**Comment by JasonLG1979** [comment]
Yep, they seem to iterate though buffer and period size combinations.
https://github.com/xbmc/xbmc/blob/a0d424ec8f1050b29c2233e57952c20f1c63c993/xbmc/cores/AudioEngine/Sinks/AESinkALSA.cpp#L758-L793

We can do the same. Basically start at a desired buffer and period size and loop though until we get something that sticks.

**Comment by roderickvd** [comment]
Good catch. Just some more lines to work around Alsa weirdness! Fondly remember the Alsa mixer rewrite. 🤭

**Comment by JasonLG1979** [comment]
@roderickvd This monstrosity seems to do the trick:

```rust
        for desired in (0..(BUFFER_TIME.as_micros() as u32)).rev() {
            match hwp.set_buffer_time_near(desired, ValueOr::Nearest) {
                Ok(actual) => {
                    if actual >= desired {
                        for desired in (0..(actual / 3)).rev() {
                            match hwp.set_period_time_near(desired, ValueOr::Nearest) {
                                Ok(actual) => {
                                    if actual >= desired {
                                        break;
                                    } else {
                                        continue;
                                    }
                                }
                                Err(_) => continue,
                            }
                        }
                    } else {
                        continue;
                    }
                }
                Err(e) => {
                    if desired == 0 {
                        return Err(SinkError::from(AlsaError::HwParams(e)));
                    } else {
                        continue;
                    }
                }
            }
        }
```

**Comment by JasonLG1979** [comment]
@jessicah try [this branch](https://github.com/JasonLG1979/librespot/tree/dynamic-alsa-buffer) and see how it works. It kinda sorta emulates what Kodi does.

**Comment by JasonLG1979** [comment]
@roderickvd That should walk though every combination of buffer and period sizes from a buffer size of 500 ms though zero with periods no larger than a 1/3 of the buffer.

The Kodi code seems to loop back though if that fails and then tries to set the period size 1st and then the buffer size. My though on that is if that's what it takes who ever wrote the damn drivers needs a kick in the butt,lol!!!

**Comment by roderickvd** [comment]
> The Kodi code seems to loop back though if that fails and then tries to set the period size 1st and then the buffer size. My though on that is if that's what it takes whoever wrote the damn drivers needs a kick in the butt,lol!!!

Either that or the hardware device (Intel?) is really iffy. What bothers me about Alsa is that it doesn't help you making these hardware idiosyncracies transparent, that you have to solve them yourself in userspace. Just blowing off steam though, it's the best thing we have in terms of bare metal performance... which means you also have to deal with bare metal.

Thanks for that patch. If it checks out with @jessicah we should vet it against various hardware configurations. Who knows, some Alsa drivers may incorrectly report MAX as 0 or other weird stuff we haven't thought of.

**Comment by JasonLG1979** [comment]
@roderickvd I can't completely emulate what Kodi is doing because some of the functions that they use are not exposed in the rust bindings yet. I may need to add them upstream 1st and it will take a while to filter down. This whole thing is a best effort approach. Hopefully "auto mode" just works with our reasonable defaults or the user may have to do some experimenting. Pretty sure why this is the reason for higher levels of abstraction like Jack, PulseAudio and PipeWire...

**Comment by JasonLG1979** [comment]
@roderickvd I believe my latest iteration comes extremely close to what Kodi does (and has done for over 9 years from the git-blame) It involves some cloning and some slide of hand though. Basically you to set the buffer size 1st and if you get what you want you set the period size. If you don't you start fresh with a previous working copy of hwp and set the periods 1st and then the buffer size. The buffer and periods are smaller but if those values have worked for them for almost 1o years they must be OK?

**Comment by JasonLG1979** [comment]
For comparison Kodi's numbers ideally get you 4 50ms periods for a total buffer of 200ms, compated to our current ideal of 5 100ms periods for a total of a 500ms buffer. The smaller buffer seems to work just fine on my Pi 4. I still need to steal my Pi Zero back from son to test on it, but even that will be less of an issue with the Pi Zero 2 that came out which is basically a Pi 3 A+ in a Zero form factor.

**Comment by JasonLG1979** [comment]
> Who knows, some Alsa drivers may incorrectly report MAX as 0 or other weird stuff we haven't thought of.

That would be extremely broken and wouldn't work at all as ALSA sits on the bottom of every Linux audio stack be it Jack, PulseAudio or PipeWire or any app or framework that sits on top of those. ALSA is the bottom of the bottom layer unless you're really dealing with real bare metal on like an embedded system like a micro controller or something.

**Comment by JasonLG1979** [comment]
Ok. The latest iteration has some useful `trace!` messages if you run in `--verbose` mode @jessicah. You should see something like this:

```
[2021-11-05T21:27:46Z TRACE librespot_playback::audio_backend::alsa] Buffer Frames: min 64, max 65536
[2021-11-05T21:27:46Z TRACE librespot_playback::audio_backend::alsa] Period Frames: min 35, max 4410
[2021-11-05T21:27:46Z TRACE librespot_playback::audio_backend::alsa] Buffer Frames: ideal 8820, actual 8820
[2021-11-05T21:27:46Z TRACE librespot_playback::audio_backend::alsa] Period Frames: ideal 2205, actual 2205
```

**Comment by JasonLG1979** [comment]
@jessicah Still waiting for you to test out [that branch](https://github.com/JasonLG1979/librespot/tree/dynamic-alsa-buffer). You're the only one that can tell us if it works on your hardware as it represents an edge case that is impossible for someone without that specific hardware to reproduce. I can confirm that it works for me on everything I own but then again the current code in the master and dev branches work just fine too so that doesn't really say anything besides that my experimental changes don't break anything on my specific hardware.

**Comment by JasonLG1979** [comment]
Just going to bump this issue one last to make sure it didn't fall though the cracks @jessicah?

@roderickvd I'd say this is very similar to https://github.com/librespot-org/librespot/issues/895 in that they both seem to be misbehaving devices, misbehaving in different ways but still misbehaving.

https://github.com/librespot-org/librespot/pull/896 would probably be about as close as we can get to "fixing" this issue also. When I say "fixing" I mean do our best and if that's not good enough use the device's default. Which is basically what Kodi does.

In any event I'd close this after whatever amount of time you decide is appropriate after this comment as @jessicah for whatever reason has not responded to requests to test possible solutions.
